### PR TITLE
ResetMustCall support

### PR DESCRIPTION
**Pull request by kelloggm**
This supersedes #280 (i.e. it includes the same functionality). It also replaces the Unconnected Socket Checker with a sound approach based on `@ResetMustCall`.

For now, the tests won't pass, because this relies on https://github.com/typetools/checker-framework/pull/4189. I doubt that will make it into the Feb 1 release (since it's been waiting for review from @mernst for days with no response, and he hasn't been responding to my messages), so we'll need to use local CF versions and test carefully rather than relying on CI for the next month (once this is merged). Sorry :(

### CODE REVIEW

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:213]: Nit: newlines are placed kind of oddly in this block comment

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:292]: Nit: newlines are placed kind of oddly in this block comment

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:679]: Is there any way to get the top-level tree for a CFG given some node in that CFG, which should be the desired `MethodTree`?  Just curious; I'm getting that getting the path for a tree might be expensive (requiring a search of the AST for the whole compilation unit)

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:726]: Should we really be crashing in this case?  Can't we just report an `incompatible.reset.mustcall` error?

**Review by msridhar — CHANGES_REQUESTED** [review]
Overall this seems great!  Just a few questions

File: build.gradle
Comment (kelloggm) [reply build.gradle:25]: No. I'm not sure why I did this - probably just a mistake. I'll undo it.

File: must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallChecker.java
Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallChecker.java:28]: I agree, but couldn't come up with a bette option. The second part of the message ("extract the unparseable expression into a local variable") is at least actionable.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (kelloggm) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:675]: There might be, but the path is needed for viewpoint adaptation later in the method regardless so computing it is unavoidable.

It might be possible to defer computing the path until after we check for the presence of an `@ResetMustCall` annotation on the declaration, if we were able to get the `MethodTree` another way - which would probably make it worthwhile.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (kelloggm) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:722]: I originally wrote this as a crash to try to figure out what else it might be, but it was never anything else besides "this" or a LocalVariable in our tests.

I think I'd prefer to leave it as a crash, since if we ever encounter that crash we'll be able to modify this method to handle that kind of receiver - that's better than silently failing, IMO.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:675]: Ok.  We don't need to block landing on this, but maybe we should add a comment if we think there may be a better way to do it in the future.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:722]: Sure, that sounds good for now

**Review by msridhar — COMMENTED** [review]
I think in terms of the changes here I am ok with them.  But before landing, we should discuss how we will stage this change alongside #291, and also whether we want to wait until the relevant Checker Framework changes land before landing this.

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (kelloggm) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:675]: I'll add a comment.

**Comment by kelloggm** [comment]
@msridhar while writing up the design of this feature for the paper, I realized that I forgot to implement one of the cases I described in my original design email: calls to ResetMustCall methods on non-owning pointers must be forbidden! Otherwise, such a call would create an untracked obligation.

I've modified the PR to cover this case, with my latest commit. I'd appreciate it if you could take a look at the new code when you get a chance.

Thanks!

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (msridhar) [inline object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java:271]: Variable names `containingMethod` and `enclosingMethod` are a bit confusing.  Maybe `enclosingMethod` and `enclosingMethodElement`?

**Review by msridhar — APPROVED** [review]
new changes look good, just one nit

File: object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java
Comment (kelloggm) [reply object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java]: Thanks for the suggestion. I also made a similar change in another part of this PR that does something similar and used the same confusing naming scheme.

**Pull request merged**
